app/src/citations.py

```diff
@@ -3,52 +3,111 @@
 import re
 from typing import Match, Sequence
 
-from src.db.models.document import Chunk, ChunkWithScore
-from src.format import _get_bem_url
+from src.db.models.document import Chunk, ChunkWithSubsection
+from src.util.bem_util import get_bem_url
 
 logger = logging.getLogger(__name__)
 
 _footnote_id = random.randint(0, 1000000)
 _footnote_index = 0
 
+CITATION_PATTERN = r"\(citation-(\d+)\)"
 
-def get_context_for_prompt(chunks: Sequence[ChunkWithScore]) -> str:
-    return "\n\n".join(
-        [
-            f"Citation: chunk-{index}\nDocument name: {chunk.chunk.document.name}{"\nHeadings: " + " > ".join(chunk.chunk.headings) if chunk.chunk.headings else ""}\nContent: {chunk.chunk.content}"
-            for index, chunk in enumerate(chunks)
-        ]
-    )
+
+def get_context(chunks: Sequence[Chunk], delimiter: str = "\n\n") -> Sequence[ChunkWithSubsection]:
+    # Given a list of chunks, split them into a flat list of subsections
+    context_mapping = []
+
+    for chunk in chunks:
+        for subsection in chunk.content.split(delimiter):
+            context_mapping.append(ChunkWithSubsection(chunk, subsection))
+
+    return context_mapping
+
+
+def get_context_for_prompt(chunks: Sequence[Chunk]) -> str:
+    context = get_context(chunks)
+
+    context_list = []
+    for index, chunk_with_subsection in enumerate(context):
+
+        if chunk_with_subsection.chunk.headings:
+            headings_text = "Headings: " + " > ".join(chunk_with_subsection.chunk.headings) + "\n"
+        else:
+            headings_text = ""
+
+        context_text = "Citation: citation-" + str(index) + "\n"
+        context_text += "Document name: " + chunk_with_subsection.chunk.document.name + "\n"
+        context_text += headings_text
+        context_text += "Content: " + chunk_with_subsection.subsection
+
+        context_list.append(context_text)
+
+    return "\n\n".join(context_list)
+
+
+def get_citation_numbers(
+    context: Sequence[ChunkWithSubsection], response: str
+) -> Sequence[ChunkWithSubsection]:
+    """
+    Map (citation-<index>) in `response`, where index is the index in `context`,
+    to a user-friendly citation number.
+    The user friendly citation number is the index of the entry in the returned list, plus one.
+    E.g., if `context` is a list with five entries, and `response` is a string like
+    "Example (citation-3)(citation-1), another example (citation-1).", then this function will return
+    [context[3], [context[1]]; citations referencing context[3] should be shown to the user as "1" and
+    citations referencing context[1] should be shown as "2".
+
+    """
+    citation_indices = re.findall(CITATION_PATTERN, response)
+
+    # Maintain order and avoid duplicates by using a list and tracking seen indices
+    seen = set()
+    citations = []
```

Comment:
Not as efficient but less code: remove `seen` and just check for `context[idx] in citations`.